pkg/frontend/clustermanager_putorpatch.go

```diff
@@ -22,24 +22,55 @@ func (f *frontend) putOrPatchClusterManagerConfiguration(w http.ResponseWriter,
 	ctx := r.Context()
 	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
 	vars := mux.Vars(r)
-	var header http.Header
-	var b []byte
+	badRequestError := fmt.Sprintf("The resource type '%s' could not be found in the namespace '%s' for api version '%s'.", vars["resourceType"], vars["resourceProviderNamespace"], vars["api-version"])
 
-	if f.apis[vars["api-version"]].ClusterManagerConfigurationConverter == nil {
-		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidResourceType, "", "The resource type '%s' could not be found in the namespace '%s' for api version '%s'.", vars["resourceType"], vars["resourceProviderNamespace"], vars["api-version"])
-		return
+	var (
+		header http.Header
+		b      []byte
+	)
+
+	switch vars["ocmResourceType"] {
+	case "syncsets":
+		if f.apis[vars["api-version"]].SyncSetConverter == nil {
+			api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidResourceType, "", badRequestError)
+			return
+		}
+	case "machinepools":
+		if f.apis[vars["api-version"]].MachinePoolConverter == nil {
+			api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidResourceType, "", badRequestError)
+			return
+		}
+	case "syncidentityproviders":
+		if f.apis[vars["api-version"]].SyncIdentityProviderConverter == nil {
+			api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidResourceType, "", badRequestError)
+			return
+		}
+	case "secrets":
+		if f.apis[vars["api-version"]].SecretConverter == nil {
+			api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidResourceType, "", badRequestError)
+			return
+		}
 	}
 
 	err := cosmosdb.RetryOnPreconditionFailed(func() error {
 		var err error
-		b, err = f._putOrPatchClusterManagerConfiguration(ctx, log, r, &header, f.apis[vars["api-version"]].ClusterManagerConfigurationConverter)
+		switch vars["ocmResourceType"] {
+		case "syncsets":
+			b, err = f._putOrPatchSyncSet(ctx, log, r, &header, f.apis[vars["api-version"]].SyncSetConverter)
+		case "machinepools":
+			b, err = f._putOrPatchMachinePool(ctx, log, r, &header, f.apis[vars["api-version"]].MachinePoolConverter)
+		case "syncidentityproviders":
+			b, err = f._putOrPatchSyncIdentityProvider(ctx, log, r, &header, f.apis[vars["api-version"]].SyncIdentityProviderConverter)
+		case "secrets":
+			b, err = f._putOrPatchSecret(ctx, log, r, &header, f.apis[vars["api-version"]].SecretConverter)
+		}
 		return err
 	})
 
 	reply(log, w, header, b, err)
 }
 
-func (f *frontend) _putOrPatchClusterManagerConfiguration(ctx context.Context, log *logrus.Entry, r *http.Request, header *http.Header, converter api.ClusterManagerConfigurationConverter) ([]byte, error) {
+func (f *frontend) _putOrPatchSyncSet(ctx context.Context, log *logrus.Entry, r *http.Request, header *http.Header, converter api.SyncSetConverter) ([]byte, error) {
```

Comment:
Currently I have separate putorpatch functions for each resource type (`_putOrPatchSyncSet`, `_putOrPatchMachinePool`, `_putOrPatchSyncIdentityProvider`, `_putOrPatchSecret`). This seems required since we now have separate converters for each resource type (`machinepool_convert.go`, `syncset_convert.go`, etc..).

Should I have kept the `clustermanager_convert.go` coverter and used it instead of having the separate putorpatch functions?

Reply:
I like having separate functions per resource type, even though it's a lot of similar code it feels cleaner.